torch/distributed/_tools/runtime_estimator.py

```diff
@@ -0,0 +1,464 @@
+                if id(e) in inp_impls:
+                    return inp_impls[id(e)]
+                else:
+                    return cls.fake_mode.fake_tensor_converter.from_real_tensor(
+                        cls.fake_mode, e
+                    )
+            else:
+                return e
+
+        return (pytree.tree_map(map_out, r), mean_op_time)
+
+    @classmethod
+    def _benchmark_estimate(cls, func, args, kwargs, res) -> float:  # type: ignore[no-untyped-def]
+        """
+        Estimates the runtime of a function using benchmarking.
+
+        Args:
+            func: The function to estimate.
+            args: The arguments to pass to the function.
+            kwargs: The keyword arguments to pass to the function.
+            res: The result of the function.
+
+        Returns:
+            float: The estimated runtime of the function in seconds.
+        """
+        assert isinstance(
+            cls.fake_mode, FakeTensorMode
+        ), "Initialize/Assign FakeTensorMode before using this function"
+        mean_op_time = 0.0
+        if func._overloadpacket not in _IGNORE_OPS:
+            try:
+                res, mean_op_time = cls._maybe_run_and_benchmark_fallback_kernel(
+                    func,
+                    args,
+                    kwargs,
+                    NotImplementedError,
+                )
+                return mean_op_time
+            except NotImplementedError:
+                cls._no_fallback_kernel.add(func._overloadpacket)
+        return mean_op_time
+
+    # Adapted from: https://github.com/pytorch/pytorch/blob/9b902b3ee3bd608a19543362b66bf06c373dd374/torch/_inductor/scheduler.py#L589  # noqa
+    @classmethod
+    def _roofline_estimate(cls, func, args, kwargs, out) -> float:  # type: ignore[no-untyped-def]
+        """
+        Estimates the runtime of a function using a roofline cost model.
+
+        Args:
+            func: The function to estimate.
+            args: The arguments to pass to the function.
+            kwargs: The keyword arguments to pass to the function.
+            out: The output of the function.
+
+        Returns:
+            float: The estimated runtime of the function in seconds.
+        """
+
+        def get_num_bytes(t: torch.Tensor) -> int:
+            st = t.untyped_storage()
+            num_bytes = st.size() * st.element_size()
+            return num_bytes
+
+        def get_compute_time(func_packet, args, kwargs, out, out_dtypes):
+            if func_packet in flop_registry:
+                assert (
+                    len(out_dtypes) == 1
+                ), f"Only support single out dtype got {out_dtypes} for {func_packet}"
+                dtype = out_dtypes.pop()
+                # We can expect to achieve 75% of theoretical peak flops
+                factor = 0.75
+                # This actually gives peta-FLOPs/s hence multiply by 1e15
+                # instead of 1e12 to get the FLOPs/s
```

Comment:
Yeah, but I have tested this function and it indeed gives peta-flops per second

Reply:
I am sure you are right since the benchmark is looking great assuming it is peta-flops. I wonder if this can be pointed out to folks that use this. In particular, do you have the documents for the test you did that shows it is peta-flops?

Reply:
I just did a manual test

```
>>> import torch
>>> from torch._inductor.utils import get_device_tflops
>>> get_device_tflops(torch.float16)
0.31186944
```
Then verified it from [NVIDIA A100 data sheet](https://www.nvidia.com/content/dam/en-zz/Solutions/Data-Center/a100/pdf/nvidia-a100-datasheet-us-nvidia-1758950-r4-web.pdf)
<img width="348" alt="Screenshot 2024-08-26 at 11 59 10 PM" src="https://github.com/user-attachments/assets/7aabc991-7530-4076-9fe5-96c73f0de4b1">

Reply:
@weifengpy @xuanzhang816 

I tested the same on H100 and it gives incorrect result, hence we are observing less accuracy on H100s.

```
>>> import torch
>>> from torch._inductor.utils import get_device_tflops
>>> get_device_tflops(torch.float16)
0.5352652800000001
>>> get_device_tflops(torch.float32)
0.033454080000000004
>>> get_device_tflops(torch.bfloat16)
0.5352652800000001
```
<img width="512" alt="Screenshot 2024-08-27 at 8 12 12 AM" src="https://github.com/user-attachments/assets/b381ba5e-db44-40b4-b499-d2bc5fbace6b">

Reply:
great finding!

Reply:
get_device_tflops is around 50% discounting at peta-flops ?

Reply:
Yes but the TensorCore FLOPS are completely different as well.